[E-Documents Core] - PINT A-NZ processing using new import process

AndriusAndrulevicius opened the pull request:
This pull request does not have a related issue as it's part of the delivery for development agreed directly with @altotovi

Implements structured format reader for PINT A-NZ e-documents to enable processing of incoming invoices and credit notes from Australian/New Zealand vendors.

Adds enum extension for draft reading capability and implements comprehensive XML parsing logic that extracts vendor information, document details, and line items from UBL-formatted documents.

Supports vendor matching through multiple identifiers including ABN, VAT registration numbers, GLN codes, and participant IDs with fallback to name/address matching.

Important

> [!IMPORTANT]
> This Pull request depends on   @#32

Fixes #

AndriusAndrulevicius committed:
Adds PINT A-NZ e-document format support for imports

Implements structured format reader for PINT A-NZ e-documents to enable processing of incoming invoices and credit notes from Australian/New Zealand vendors.

Adds enum extension for draft reading capability and implements comprehensive XML parsing logic that extracts vendor information, document details, and line items from UBL-formatted documents.

Supports vendor matching through multiple identifiers including ABN, VAT registration numbers, GLN codes, and participant IDs with fallback to name/address matching.

Updates version to 27.0.37120.0 to reflect new functionality.

AndriusAndrulevicius committed:
Resets version to 27.0.0.0

Standardizes the version number format by removing the build number suffix, likely preparing for a new release cycle or aligning with versioning conventions.

tinestaric commented on Apps/APAC/EDocumentFormats/PINT A-NZ/app/src/Core/PINTANZReadDraftHandler.Codeunit.al:
```diff
@@ -0,0 +1,279 @@
+    var
+        EDocumentImportHelper: Codeunit "E-Document Import Helper";
+
+    procedure ReadIntoDraft(EDocument: Record "E-Document"; TempBlob: Codeunit "Temp Blob"): Enum "E-Doc. Process Draft"
```
xml docs

tinestaric commented on Apps/APAC/EDocumentFormats/PINT A-NZ/app/src/Core/PINTANZReadDraftHandler.Codeunit.al:
```diff
@@ -0,0 +1,279 @@
+        EDocumentPurchaseHeader.InsertForEDocument(EDocument);
+
+        TempBlob.CreateInStream(DocStream, TextEncoding::UTF8);
+        XmlDocument.ReadFrom(DocStream, PINTANZXml);
```
nit: stream var could be dropped

tinestaric commented on Apps/APAC/EDocumentFormats/PINT A-NZ/app/src/Core/PINTANZReadDraftHandler.Codeunit.al:
```diff
@@ -0,0 +1,279 @@
+                PopulateEDocumentForCreditNote(PINTANZXml, XmlNamespaces, EDocumentPurchaseHeader, EDocument);
+        end;
+
+        EDocumentPurchaseHeader.Modify();
```
implicit trigger

tinestaric commented on Apps/APAC/EDocumentFormats/PINT A-NZ/app/src/Core/PINTANZReadDraftHandler.Codeunit.al:
```diff
@@ -0,0 +1,279 @@
+        LineXMLList: XmlNodeList;
+        LineXMLNode: XmlNode;
+        i: Integer;
+        InvoiceLinePathLbl: Label '/inv:Invoice/cac:InvoiceLine';
```
how come you've extracted these two while in other places they're hardcoded?

AndriusAndrulevicius replied:
AI 🤷‍♂️

tinestaric commented on Apps/APAC/EDocumentFormats/PINT A-NZ/app/src/Core/PINTANZReadDraftHandler.Codeunit.al:
```diff
@@ -0,0 +1,279 @@
+            NewLineXML := XmlDocument.Create();
+            NewLineXML.Add(LineXMLNode.AsXmlElement());
+            PopulatePINTANZPurchaseLine(NewLineXML, XmlNamespaces, EDocumentPurchaseLine);
+            EDocumentPurchaseLine.Insert();
```
implicit trigger

tinestaric commented on Apps/APAC/EDocumentFormats/PINT A-NZ/app/src/Core/PINTANZReadDraftHandler.Codeunit.al:
```diff
@@ -0,0 +1,279 @@
+        if InputABN = '' then
+            exit(false);
+
+        Vendor.SetRange(ABN, InputABN);
```
Honestly, I'm not used to using it myself, but I think all these simple finders could all have SetLoadFields

AndriusAndrulevicius committed:
Refactors PINT A-NZ handler with documentation and optimizations

Renames codeunit to better reflect its purpose as a comprehensive handler rather than just a draft reader.

Adds comprehensive XML documentation for public procedures to improve code maintainability and developer experience.

Optimizes XML processing by eliminating intermediate stream variable and using direct method chaining.

Removes unused namespace import and redundant local helper method by leveraging existing XML helper functionality.

Improves database performance by disabling trigger execution on non-critical insert and modify operations and adding field loading optimization for vendor lookups.

AndriusAndrulevicius committed:
merge main into dev/aan/PINT_A-NZ-v2

GMatuleviciute commented on Apps/APAC/EDocumentFormats/PINT A-NZ/app/src/Core/EDocumentPINTANZHandler.Codeunit.al:
```diff
@@ -0,0 +1,281 @@
+namespace Microsoft.eServices.EDocument.Format;
+
+using Microsoft.eServices.EDocument.Processing.Interfaces;
+using Microsoft.eServices.EDocument.Processing.Import;
+using Microsoft.eServices.EDocument.Processing.Import.Purchase;
+using Microsoft.eServices.EDocument.Service.Participant;
+using Microsoft.eServices.EDocument.Helpers;
+using Microsoft.Purchases.Vendor;
+using System.Utilities;
+using Microsoft.eServices.EDocument;
```
Please sort usings

GMatuleviciute commented on Apps/APAC/EDocumentFormats/PINT A-NZ/app/src/Core/EDocumentPINTANZHandler.Codeunit.al:
```diff
@@ -0,0 +1,281 @@
+            EDocumentPurchaseHeader."[BC] Vendor No." := VendorNo;
+        InsertPINTANZPurchaseInvoiceLines(PINTANZXml, XmlNamespaces, EDocumentPurchaseHeader."E-Document Entry No.");
+    end;
+#pragma warning disable AA0139 // false positive: overflow handled by SetStringValueInField
```
```suggestion
#pragma warning disable AA0139 // false positive: overflow handled by SetStringValueInField

```

AndriusAndrulevicius committed:
Update Apps/APAC/EDocumentFormats/PINT A-NZ/app/src/Core/EDocumentPINTANZHandler.Codeunit.al

Co-authored-by: Grasiele Matuleviciute <131970463+GMatuleviciute@users.noreply.github.com>

AndriusAndrulevicius committed:
Refactors constants and improves loop efficiency

Moves GLN scheme identifier constant to class level for better organization and reusability across methods.

Changes label constants to token constants with Locked attribute for better maintainability and localization handling.

Replaces indexed for loop with foreach loop for better performance and readability when processing XML line nodes.

Reorganizes using statements in alphabetical order for improved code organization.

GMatuleviciute commented on Apps/APAC/EDocumentFormats/PINT A-NZ/app/src/Core/EDocumentPINTANZHandler.Codeunit.al:
```diff
@@ -0,0 +1,284 @@
+    var
+        EDocumentImportHelper: Codeunit "E-Document Import Helper";
+        GLNSchemeIdTok: Label '0088', Locked = true;
+
```
```suggestion
```

GMatuleviciute commented on Apps/APAC/EDocumentFormats/PINT A-NZ/app/src/Core/EDocumentPINTANZHandler.Codeunit.al:
```diff
@@ -0,0 +1,284 @@
+                GLN := CopyStr(EDocumentXMLHelper.GetNodeValue(PINTANZXml, XmlNamespaces, BasePathTxt + '/cbc:EndpointID'), 1, MaxStrLen(GLN));
+                EDocumentPurchaseHeader."Vendor GLN" := GLN;
+            end;
+
```
```suggestion
```

GMatuleviciute committed:
Update Apps/APAC/EDocumentFormats/PINT A-NZ/app/src/Core/EDocumentPINTANZHandler.Codeunit.al

GMatuleviciute committed:
Update Apps/APAC/EDocumentFormats/PINT A-NZ/app/src/Core/EDocumentPINTANZHandler.Codeunit.al

AndriusAndrulevicius committed:
Sets document type for PINT A-NZ purchase documents

Ensures purchase invoices and credit memos are properly classified by setting the E-Document Type field during document processing.

This prevents potential issues with document categorization and improves data consistency in the PINT A-NZ electronic document handling workflow.

GMatuleviciute committed:
Apply suggestion from @GMatuleviciute

AndriusAndrulevicius committed:
Merge remote-tracking branch 'origin/main' into dev/aan/PINT_A-NZ-v2

AndriusAndrulevicius committed:
Add PINT A‑NZ structured tests and draft reset

Introduces integration tests for the PINT A‑NZ structured import flow that validate draft extraction, viewing extracted data, and finalizing into a purchase document using embedded sample data.

Adds a mock e-document format and a test helper library to simplify setup and assertions, extends test configuration (ID range, OnPrem target, resource folder), and declares a dependency on the PINT A‑NZ tests package.

Adds a draft reset procedure to remove draft purchase data for an e-document, ensuring clean re-runs and predictable test states.

AndriusAndrulevicius committed:
missing test file added

AndriusAndrulevicius committed:
Improves PINT A‑NZ tests; drops OnPrem target

Switches the test file path to a locked label to avoid hard-coded strings and ease maintenance.
Prevents trigger execution during setup/cleanup for faster, more deterministic tests.
Adds BDD-style comments to clarify test intent.
Removes the OnPrem target from the test app manifest to align with cloud environments.
